### ISSUE TITLE
Switch Redux logic setup to use Redux Toolkit (take 2)

### ISSUE THREAD

**Issue by markerikson** [react-boilerplate/react-boilerplate#2806]
**Is your feature request related to a problem? Please describe.**
Hi, I'm a Redux maintainer. I'd like to offer a couple suggestions.

The current Redux example implementation shows use of reducers with "hand-written" immutable update logic, and has a somewhat complex store configuration.  In addition, the project's current "feature folder" structure results in multiple separate files for the Redux logic in each feature.

**Describe the solution you'd like**

The project should switch to using the official [Redux Toolkit package](https://redux-toolkit.js.org) to implement the Redux logic.

The project should also consider switching to use the "ducks" pattern within each feature folder.

**Describe alternatives you've considered**
N/A

**Additional context**

Redux Toolkit is our new official, opinionated, batteries-included toolset for efficient Redux development.  It includes utilities to simplify several common Redux use cases, including store setup, defining reducers, immutable update logic, and even creating entire "slices" of state at once without writing any action creators or action types by hand.

We are recommending RTK as the default standard way to write Redux logic.  Partly related to that, we are also recommending that developers should prefer to use "feature folders" or "ducks", and RTK makes it particularly easy to write "ducks"-structured Redux code.  (Reference: the work-in-progress [Redux "Style Guide" docs page](https://deploy-preview-3626--redux-docs.netlify.com/style-guide/style-guide), and [the RTK "Usage Guide" page](https://redux-toolkit.js.org/usage/usage-guide#exporting-and-using-slices).)

Since many developers will be starting their projects using `react-boilerplate`, it would be beneficial if Redux Toolkit was included as part of the default project setup, as it will greatly simplify the application code they write.

**Comment by markerikson** [comment]
As an additional observation: the current use of Reselect borders on being wrong.  In particular:

```js
// selectors.js
const makeSelectCurrentUser = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.currentUser,
  );

// elsewhere
const stateSelector = createStructuredSelector({
  currentUser: makeSelectCurrentUser(),
});
```

The only time you need to create multiple instances of a selector is when [it will be called with multiple varying inputs that cause memoization to fail](https://blog.isquaredsoftware.com/2017/12/idiomatic-redux-using-reselect-selectors/#advanced-optimizations-with-react-redux), such as list items that select their own data item from state by ID.

There's no need to have multiple instances of a `selectCurrentUser` selector, because it's always retrieving exactly the same piece of state, with only one input: `state`.  

So, it should really just be:

```js
const selectCurrentUser = createSelector(
    [selectGlobal],
    globalState => globalState.currentUser,
);
```

Sorta-similarly, this is also unnecessary:

```js
const stateSelector = createStructuredSelector({
  currentUser: makeSelectCurrentUser(),
});

export default function RepoListItem({ item }) {
  const { currentUser } = useSelector(stateSelector);
```

If all you care about is the `currentUser` value, there's no reason to creat a wrapper selector that puts it in an object, only to destructure it right back out.  Just do:

```js
import {selectCurrentUser} from "containers/App/selectors";

export default function RepoListItem({ item }) {
  const currentUser = useSelector(selectCurrentUser);
```

**Comment by julienben** [comment]
Hey @markerikson, Thanks a lot for your help and suggestions here! I'm very excited to see this as I was planning on doing this migration anyway and wasn't expecting you to come and open a PR yourself.

I'll review the PR directly but wanted to answer everything else here.

**Regarding switching to RTK and the ducks pattern:** 100% down with this change. We've been looking for a way to simplify the way we write Redux code for a while. (See [this comment](https://github.com/react-boilerplate/react-boilerplate/issues/2725#issuecomment-526240327) for example.)

**Regarding your mention of feature folders**: Are you suggesting we move away from the containers/components pattern? We're certainly not married to it. We mostly haven't switched because we haven't seen a suggestion for a pattern with clear advantages over it. (It would also be a large project given that our generators and much of our documentation is built around it.)

**Regarding selectors:** "Borders on being wrong" is a kind euphemism. 😅 Thank you for the detailed feedback! I'll plan a PR.

We indeed have many developers starting their projects with react-boilerplate and thus have a responsibility to promote the best practices in the industry.

**Comment by markerikson** [comment]
The "container folders" you have are fairly close to being "feature folders" already, as opposed to having separate folders like `/actions` and `/reducers`.  In this case, it's more that each folder still has multiple files for each type of thing.  To be clear, that's not "wrong" at all, but it does illustrate the common complaint that "using Redux requires that I touch a dozen files just to make one feature work!".

Also, looking at the linked comment... it's amazing how close `createSlice()` comes to `magicalFunctionOfMyDreams()` :)

**Comment by julienben** [comment]
I know... You actually wrote it, the magical function of my dreams 😂

My name is better though.

**Comment by vineyardbovines** [comment]
I'll echo what @julienben said, along with the fact that I love the ducks pattern and use it everyday, hell yeah I'd love to use it here too.

**Comment by BenLorantfy** [comment]
I haven't measured it but I think the "selector factory" pattern is also pretty bad for performance. It uses more memory than needed for the extra caches and uses more CPU than needed since the cache invalidation check needs to be run more often.

**Comment by markerikson** [comment]
Realistically, most of those extra selectors are just getting thrown away and garbage collected

**Comment by BenLorantfy** [comment]
I definitely think feature folders / ducks is better than doing per-page redux. At work we're using this pattern:
```
import { useInjectReducer, useInjectSaga } from "redux-injectors";

const UsersManager = function() {
  useInjectReducer({ key: "users", reducer: usersReducer });
  useInjectSaga({ key: "users", saga: usersSaga });

  return null; // Just return null. Intentionally doesn't return any UI here
}
```

With this folder structure
```
/containers
/components
/pages
/managers
  /UsersManager
    /reducer.js
    /selectors.js
    /saga.js
    /index.js
```

So our redux isn't coupled to any one page or UI, but we still get the code-splitting benefits of `UsersManager`. We just render `<UsersManager />` on any page that needs the users state.

**Comment by markerikson** [comment]
Yeah, I don't want to dictate the exact form of the folder structure you or other folks use.  I just want to encourage folks to _not_ use "folder-by-type", and hopefully saying "try ducks" will cut down on the complaints about Redux "requiring too many separate files" :)

**Comment by Mujaddadi** [comment]
I second the folder structure change and the complexity complex store configuration. At the moment the folder structure is not really scalable as we are putting everything in either components or containers. If you have a small app that will work but if you are working on very big app like me, these two folders will become huge. 

I have used this boilerplate in couple of projects and every time I had to to change the folder structure to feature based instead of just two folders structure even though I was not aware about what is a "duck" pattern (It seems similar). So we put the container and related components in on feature folder and anything shared into shared component folder.

**Comment by Mujaddadi** [comment]
It turns out what we followed for the past three years is more closer to something now called [re-duck](https://www.freecodecamp.org/news/scaling-your-redux-app-with-ducks-6115955638be/) structure. Could be better option for this boiler plate.

**Comment by Mujaddadi** [comment]
This is how our folder structure looks like for one feature and it works quite good for us. 
![Folder](https://user-images.githubusercontent.com/2556882/70426276-67067f00-1a7b-11ea-8dad-230777b5816a.png)

**Comment by joejordan** [comment]
Just a personal opinion, but part of the appeal for me in adopting React Boilerplate over others was the comparative lack of "magic" that the framework handled, combined with what appeared to me as the implementation of industry best practices in organization and code.

In that spirit, I'm not sure I agree with the approach of incorporating another library that obfuscates some of Redux's implementation details. I personally like seeing how a store is created and configured with bare Redux, as it offers both a learning opportunity and the ability to customize to one's heart's content.

I am less familiar with the "ducks" pattern discussed here, but the existing pattern/layout of action.js, reducer.js and optionally saga.js with a well-identified constants.js to define how the action appears helps to clearly delineate the mechanisms inherent in Redux.

**Comment by markerikson** [comment]
@joejordan : part of the point of RTK is that it _doesn't_ obfuscate the fact that you're using Redux:

https://blog.isquaredsoftware.com/2019/10/redux-starter-kit-1.0/#defining-a-roadmap-and-clarifying-the-vision

You're still writing reducers and dispatching actions, and setting up a store that has a reducer function and some middleware.  You just don't have to do all that work by hand any more.  Besides, those aspects are "incidental complexity" - they're not a relevant or useful part of what defines Redux.  

Similarly, the "ducks pattern" reduces the incidental complexity by minimizing the number of files you have to touch.

Now, note that adding RTK as the default example here doesn't prevent you from doing everything by hand if you really want to.  It _does_ simplify the setup and point you in an easier way of doing things by default.

And, as mentioned at the top, we are officially recommending that [people should use RTK as the standard way to write Redux logic](https://redux.js.org/style-guide/style-guide#use-redux-toolkit-for-writing-redux-logic), and [organize files as "feature folders" or "ducks"](https://redux.js.org/style-guide/style-guide#structure-files-as-feature-folders-or-ducks).  You're always welcome to do something different if you want, but my goal _is_ to make those the "industry standard".

**Comment by lsanders** [comment]
@Can-Sahin, curious to understand how this impacts react-boilerplate-typescript. I didn't look deeply at it, but wonder if we lose some of the type-safety with the action creators since they take payloads now? Do you know when you'll tackle the merge? I know that's diverging quickly these days, but wanted to say that I absolutely love the typescript fork. Also, happy to help, if you think you can carve out anything.

**Comment by markerikson** [comment]
@lsanders : no, Redux Toolkit is written in TypeScript itself, and we have specifically designed its API to minimize the amount of explicit type declarations you actually have to write yourself:

https://redux-toolkit.js.org/tutorials/advanced-tutorial

https://redux-toolkit.js.org/usage/usage-with-typescript
